request-example.go

```diff
@@ -69,6 +69,28 @@ func (fs *root) Filewrite(r *Request) (io.WriterAt, error) {
 	return file.WriterAt()
 }
 
+func (fs *root) OpenFile(r *Request) (WriterAtReaderAt, error) {
+	if fs.mockErr != nil {
+		return nil, fs.mockErr
+	}
+	_ = r.WithContext(r.Context()) // initialize context for deadlock testing
+	fs.filesLock.Lock()
+	defer fs.filesLock.Unlock()
+	file, err := fs.fetch(r.Filepath)
+	if err == os.ErrNotExist {
+		dir, err := fs.fetch(filepath.Dir(r.Filepath))
+		if err != nil {
+			return nil, err
+		}
+		if !dir.isdir {
+			return nil, os.ErrInvalid
+		}
+		file = newMemFile(r.Filepath, false)
```

Comment:
Hm, we control this function, and it’s not exported. So we might be able to refactor it to return an `error` as well, and we can then put the `fs.fetch(filepath.Dir(…))` and `!dir.isdir` checks into that function? As well, it could be a receiver method on the `fs` struct, and assign the `fs.files[filepath] = file` as well?

Also, I’m not sure if we want `filepath` here, do we? We’re working with memory files, which means we should be able to use the POSIX abstraction regardless of underlying OS. But I guess if the other code is already using `filepath`, then no. 😆